File: src/graph/go-builder.js
import { basename } from "node:path";
import { readText, walkFiles } from "../core/files.js";
import { addEdge, addNode, createGraph, makeNodeId, renderGraphReport } from "./graph.js";

const CALL_EXCLUDES = new Set([
  "if",
  "for",
  "switch",
  "return",
  "range",
  "go",
  "defer",
  "select",
  "append",
  "make",
  "new",
  "len",
  "cap",
  "copy",
  "delete",
  "panic",
  "recover"
]);

export async function buildGoGraph(repoRoot) {
  const graph = createGraph({ repoRoot, source: "go-static-fallback" });
  const files = walkFiles(repoRoot, {
    extensions: [".go", ".md", ".txt", ".yaml", ".yml"]
  });
  const symbols = new Map();

  for (const file of files) {
    const content = await readText(file.path);
    const fileId = makeNodeId("file", [file.relativePath]);
    addNode(graph, {
      id: fileId,
      type: "file",
      name: basename(file.relativePath),
      path: file.relativePath,
      metadata: {
        bytes: file.size
      }
    });

    if (file.relativePath.endsWith(".go")) {
      extractGoFile(graph, symbols, file, content, fileId);
    } else {
      extractDocFile(graph, file, content, fileId);
    }
  }

  connectCalls(graph, symbols);
  graph.report = renderGraphReport(graph);
  graph.stats = {
    files: files.length
  };
  return graph;
}

function extractDocFile(graph, file, content, fileId) {
  const headingPattern = /^(#{1,6})\s+(.+)$/gm;
  for (const match of content.matchAll(headingPattern)) {
    const title = match[2].trim();
    const line = lineNumberAt(content, match.index);
    const nodeId = makeNodeId("doc-heading", [file.relativePath, title]);
    addNode(graph, {
      id: nodeId,
      type: "doc-heading",
      name: title,
      path: file.relativePath,
      line,
      metadata: {
        level: match[1].length
      }
    });
    addEdge(graph, { source: fileId, target: nodeId, type: "contains" });
  }
}

function extractGoFile(graph, symbols, file, content, fileId) {
  const packageMatch = content.match(/^package\s+([A-Za-z_][A-Za-z0-9_]*)/m);
  const packageName = packageMatch?.[1] ?? "unknown";
  const packageId = makeNodeId("package", [packageName]);
  addNode(graph, {
    id: packageId,
    type: "package",
    name: packageName,
    metadata: {}
  });
  addEdge(graph, { source: fileId, target: packageId, type: "declares_package" });

  const imports = extractImports(content);
  for (const importPath of imports) {
    const importId = makeNodeId("import", [importPath]);
    addNode(graph, {
      id: importId,
      type: "import",
      name: importPath,
      metadata: {}
    });
    addEdge(graph, { source: fileId, target: importId, type: "imports" });
  }

  const functionPattern = /^func\s+(?:\(([^)]*)\)\s*)?([A-Za-z_][A-Za-z0-9_]*)\s*\(/gm;
  for (const match of content.matchAll(functionPattern)) {
    const receiver = normaliseReceiver(match[1]);
    const name = match[2];
    const line = lineNumberAt(content, match.index);
    const qualified = receiver ? `${receiver}.${name}` : name;
    const nodeId = makeNodeId("function", [file.relativePath, qualified]);
    const body = extractFunctionBody(content, match.index);
    const calls = extractCalls(body);
    addNode(graph, {
      id: nodeId,
      type: file.relativePath.endsWith("_test.go") || name.startsWith("Test") ? "test" : "function",
      name: qualified,
      path: file.relativePath,
      line,
      metadata: {
        package: packageName,
        receiver,
        calls
      }
    });
    symbols.set(name, nodeId);
    symbols.set(qualified, nodeId);
    addEdge(graph, { source: fileId, target: nodeId, type: "contains" });
    addEdge(graph, { source: nodeId, target: packageId, type: "belongs_to" });
  }

  const mapEntryPattern = /^\s*"([^"]+)"\s*:\s*([A-Za-z_][A-Za-z0-9_]*),/gm;
  for (const match of content.matchAll(mapEntryPattern)) {
    const tag = match[1];
    const fn = match[2];
    if (!looksLikeValidatorTag(tag)) {
      continue;
    }
    const line = lineNumberAt(content, match.index);
    const tagId = makeNodeId("validator-tag", [tag]);
    addNode(graph, {
      id: tagId,
      type: "validator-tag",
      name: tag,
      path: file.relativePath,
      line,
      metadata: {
        implementation: fn
      }
    });
    addEdge(graph, { source: fileId, target: tagId, type: "contains" });
    addEdge(graph, { source: tagId, target: makeNodeId("symbol-ref", [fn]), type: "implemented_by" });
  }

  const regexPattern = /^var\s+([A-Za-z_][A-Za-z0-9_]*)\s*=\s*regexp\.MustCompile\(/gm;
  for (const match of content.matchAll(regexPattern)) {
    const nodeId = makeNodeId("regex", [match[1]]);
    addNode(graph, {
      id: nodeId,
      type: "regex",
      name: match[1],
      path: file.relativePath,
      line: lineNumberAt(content, match.index),
      metadata: {}
    });
    addEdge(graph, { source: fileId, target: nodeId, type: "contains" });
  }
}

function connectCalls(graph, symbols) {
  const callableNodes = graph.nodes.filter((node) => node.metadata?.calls?.length);
  for (const node of callableNodes) {
    for (const call of node.metadata.calls) {
      const target = symbols.get(call);
      if (target) {
        addEdge(graph, { source: node.id, target, type: "calls" });
      }
    }
  }

  for (const edge of graph.edges.filter((item) => item.type === "implemented_by")) {
    const symbol = edge.target.replace(/^symbol-ref:/, "");
    const target = symbols.get(symbol);
    if (target) {
      edge.target = target;
    } else {
      addNode(graph, {
        id: edge.target,
        type: "symbol-ref",
        name: symbol,
        metadata: {}
      });
    }
  }
}

function extractImports(content) {
  const imports = [];
  const blockPattern = /import\s*\(([\s\S]*?)\)/m;
  const block = content.match(blockPattern);
  if (block) {
    for (const match of block[1].matchAll(/"([^"]+)"/g)) {
      imports.push(match[1]);
    }
  }
  for (const match of content.matchAll(/^import\s+"([^"]+)"/gm)) {
    imports.push(match[1]);
  }
  return [...new Set(imports)];
}

function extractCalls(body) {
  const calls = new Set();
  for (const match of body.matchAll(/\b([A-Za-z_][A-Za-z0-9_]*)\s*\(/g)) {
    const name = match[1];
    if (!CALL_EXCLUDES.has(name)) {
      calls.add(name);
    }
  }
  return [...calls];
}

function extractFunctionBody(content, start) {
  const braceStart = content.indexOf("{", start);
  if (braceStart < 0) {
    return "";
  }
  let depth = 0;
  for (let index = braceStart; index < content.length; index += 1) {
    if (content[index] === "{") {
      depth += 1;
    } else if (content[index] === "}") {
      depth -= 1;
      if (depth === 0) {
        return content.slice(braceStart, index + 1);
      }
    }
  }
  return content.slice(braceStart);
}

function normaliseReceiver(receiver) {
  if (!receiver) {
    return null;
  }
  const cleaned = receiver.replace(/\*/g, "").trim();
  const parts = cleaned.split(/\s+/);
  return parts[parts.length - 1] || null;
}

function lineNumberAt(content, index) {
  return content.slice(0, index).split("\n").length;
}

function looksLikeValidatorTag(tag) {
  return /^[a-z][a-z0-9_|\-.]+$/i.test(tag) && tag.length <= 80;
}

File: go_issue_agent/graph/go_builder.py
from __future__ import annotations

import re
from pathlib import Path

from go_issue_agent.core.files import read_text, walk_files
from go_issue_agent.graph.graph import add_edge, add_node, create_graph, make_node_id, render_graph_report


CALL_EXCLUDES = {
    "if",
    "for",
    "switch",
    "return",
    "range",
    "go",
    "defer",
    "select",
    "append",
    "make",
    "new",
    "len",
    "cap",
    "copy",
    "delete",
    "panic",
    "recover",
}


def build_go_graph(repo_root: str) -> dict:
    graph = create_graph(repo_root, source="go-static-fallback")
    files = walk_files(repo_root, extensions=[".go", ".md", ".txt", ".yaml", ".yml"])
    symbols: dict[str, str] = {}

    for file in files:
        content = read_text(file["path"])
        file_id = make_node_id("file", [file["relativePath"]])
        add_node(
            graph,
            {
                "id": file_id,
                "type": "file",
                "name": Path(file["relativePath"]).name,
                "path": file["relativePath"],
                "metadata": {"bytes": file["size"]},
            },
        )
        if file["relativePath"].endswith(".go"):
            _extract_go_file(graph, symbols, file, content, file_id)
        else:
            _extract_doc_file(graph, file, content, file_id)

    _connect_calls(graph, symbols)
    graph["report"] = render_graph_report(graph)
    graph["stats"] = {"files": len(files)}
    return graph


def _extract_doc_file(graph: dict, file: dict, content: str, file_id: str) -> None:
    for match in re.finditer(r"^(#{1,6})\s+(.+)$", content, re.MULTILINE):
        title = match.group(2).strip()
        line = _line_number_at(content, match.start())
        node_id = make_node_id("doc-heading", [file["relativePath"], title])
        add_node(
            graph,
            {
                "id": node_id,
                "type": "doc-heading",
                "name": title,
                "path": file["relativePath"],
                "line": line,
                "metadata": {"level": len(match.group(1))},
            },
        )
        add_edge(graph, {"source": file_id, "target": node_id, "type": "contains"})


def _extract_go_file(graph: dict, symbols: dict[str, str], file: dict, content: str, file_id: str) -> None:
    package_match = re.search(r"^package\s+([A-Za-z_][A-Za-z0-9_]*)", content, re.MULTILINE)
    package_name = package_match.group(1) if package_match else "unknown"
    package_id = make_node_id("package", [package_name])
    add_node(graph, {"id": package_id, "type": "package", "name": package_name, "metadata": {}})
    add_edge(graph, {"source": file_id, "target": package_id, "type": "declares_package"})

    for import_path in _extract_imports(content):
        import_id = make_node_id("import", [import_path])
        add_node(graph, {"id": import_id, "type": "import", "name": import_path, "metadata": {}})
        add_edge(graph, {"source": file_id, "target": import_id, "type": "imports"})

    for match in re.finditer(r"^func\s+(?:\(([^)]*)\)\s*)?([A-Za-z_][A-Za-z0-9_]*)\s*\(", content, re.MULTILINE):
        receiver = _normalise_receiver(match.group(1))
        name = match.group(2)
        line = _line_number_at(content, match.start())
        qualified = f"{receiver}.{name}" if receiver else name
        node_id = make_node_id("function", [file["relativePath"], qualified])
        body = _extract_function_body(content, match.start())
        calls = _extract_calls(body)
        node_type = "test" if file["relativePath"].endswith("_test.go") or name.startswith("Test") else "function"
        add_node(
            graph,
            {
                "id": node_id,
                "type": node_type,
                "name": qualified,
                "path": file["relativePath"],
                "line": line,
                "metadata": {"package": package_name, "receiver": receiver, "calls": calls},
            },
        )
        symbols[name] = node_id
        symbols[qualified] = node_id
        add_edge(graph, {"source": file_id, "target": node_id, "type": "contains"})
        add_edge(graph, {"source": node_id, "target": package_id, "type": "belongs_to"})

    for match in re.finditer(r'^\s*"([^"]+)"\s*:\s*([A-Za-z_][A-Za-z0-9_]*),', content, re.MULTILINE):
        tag = match.group(1)
        implementation = match.group(2)
        if not _looks_like_validator_tag(tag):
            continue
        tag_id = make_node_id("validator-tag", [tag])
        add_node(
            graph,
            {
                "id": tag_id,
                "type": "validator-tag",
                "name": tag,
                "path": file["relativePath"],
                "line": _line_number_at(content, match.start()),
                "metadata": {"implementation": implementation},
            },
        )
        add_edge(graph, {"source": file_id, "target": tag_id, "type": "contains"})
        add_edge(graph, {"source": tag_id, "target": make_node_id("symbol-ref", [implementation]), "type": "implemented_by"})

    for match in re.finditer(r"^var\s+([A-Za-z_][A-Za-z0-9_]*)\s*=\s*regexp\.MustCompile\(", content, re.MULTILINE):
        regex_id = make_node_id("regex", [match.group(1)])
        add_node(
            graph,
            {
                "id": regex_id,
                "type": "regex",
                "name": match.group(1),
                "path": file["relativePath"],
                "line": _line_number_at(content, match.start()),
                "metadata": {},
            },
        )
        add_edge(graph, {"source": file_id, "target": regex_id, "type": "contains"})


def _connect_calls(graph: dict, symbols: dict[str, str]) -> None:
    callable_nodes = [node for node in graph["nodes"] if node.get("metadata", {}).get("calls")]
    for node in callable_nodes:
        for call in node["metadata"]["calls"]:
            target = symbols.get(call)
            if target:
                add_edge(graph, {"source": node["id"], "target": target, "type": "calls"})

    for edge in [item for item in graph["edges"] if item["type"] == "implemented_by"]:
        symbol = edge["target"].replace("symbol-ref:", "", 1)
        target = symbols.get(symbol)
        if target:
            edge["target"] = target
        else:
            add_node(graph, {"id": edge["target"], "type": "symbol-ref", "name": symbol, "metadata": {}})


def _extract_imports(content: str) -> list[str]:
    imports = set()
    block = re.search(r"import\s*\(([\s\S]*?)\)", content, re.MULTILINE)
    if block:
        for match in re.finditer(r'"([^"]+)"', block.group(1)):
            imports.add(match.group(1))
    for match in re.finditer(r'^import\s+"([^"]+)"', content, re.MULTILINE):
        imports.add(match.group(1))
    return sorted(imports)


def _extract_calls(body: str) -> list[str]:
    calls = set()
    for match in re.finditer(r"\b([A-Za-z_][A-Za-z0-9_]*)\s*\(", body):
        name = match.group(1)
        if name not in CALL_EXCLUDES:
            calls.add(name)
    return sorted(calls)


def _extract_function_body(content: str, start: int) -> str:
    brace_start = content.find("{", start)
    if brace_start < 0:
        return ""
    depth = 0
    for index in range(brace_start, len(content)):
        if content[index] == "{":
            depth += 1
        elif content[index] == "}":
            depth -= 1
            if depth == 0:
                return content[brace_start : index + 1]
    return content[brace_start:]


def _normalise_receiver(receiver: str | None) -> str | None:
    if not receiver:
        return None
    cleaned = receiver.replace("*", "").strip()
    parts = re.split(r"\s+", cleaned)
    return parts[-1] if parts else None


def _line_number_at(content: str, index: int) -> int:
    return content[:index].count("\n") + 1


def _looks_like_validator_tag(tag: str) -> bool:
    return bool(re.match(r"^[a-z][a-z0-9_|\-.]+$", tag, re.IGNORECASE)) and len(tag) <= 80
